frmts/vrt/vrtprocesseddatasetfunctions.cpp

```diff
@@ -1468,6 +1471,259 @@ static CPLErr TrimmingProcess(
+            return CE_Failure;
+        }
+
+        m_adfValuesForPixel.resize(m_nCurrentBatchSize);
+
+        for (int i = 0; i < m_nCurrentBatchSize; i++)
+        {
+            std::string osVar = "B" + std::to_string(i + 1);
+            m_oExpression->RegisterVariable(osVar, &m_adfValuesForPixel[i]);
+        }
+
+        // TODO only register "BANDS" if it is used
+        m_oExpression->RegisterVector("BANDS", &m_adfValuesForPixel);
+
+        return m_oExpression->Compile();
+    }
+
+    CPLErr Evaluate(const double *padfInputs, size_t nExpectedOutBands)
+    {
+        m_adfResults.clear();
+
+        for (int iBatch = 0; iBatch < BatchCount(); iBatch++)
+        {
+            const auto nBandsRemaining =
+                static_cast<int>(m_nInBands - (m_nNominalBatchSize * iBatch));
+            const auto nBatchSize =
+                std::min(m_nNominalBatchSize, nBandsRemaining);
+
+            // FIXME this is overly general - we really should have a maximum
+            // of two possible batch sizes. If we have two batch sizes, we should
+            // keep two expression objects around to avoid reparsing the expression
+            // every time the batch size changes (every pixel)
+            if (m_nCurrentBatchSize != nBatchSize)
+            {
+                m_nCurrentBatchSize = nBatchSize;
+                if (auto eErr = Compile(); eErr != CE_None)
+                {
+                    return eErr;
+                }
+            }
+
+            std::copy(padfInputs + iBatch * (m_nNominalBatchSize),
+                      padfInputs + (iBatch + 1) * m_nNominalBatchSize,
+                      m_adfValuesForPixel.begin());
+
+            if (auto eErr = m_oExpression->Evaluate(); eErr != CE_None)
+            {
+                return eErr;
+            }
+
+            const auto &adfResults = m_oExpression->Results();
+            if (BatchCount() > 1)
+            {
+                std::copy(adfResults.begin(), adfResults.end(),
+                          std::back_inserter(m_adfResults));
+            }
+        }
+
+        if (nExpectedOutBands > 0)
+        {
+            if (Results().size() != static_cast<std::size_t>(nExpectedOutBands))
+            {
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Expression returned %d values but "
+                         "%d output bands were expected.",
+                         static_cast<int>(Results().size()),
+                         static_cast<int>(nExpectedOutBands));
+                return CE_Failure;
+            }
+        }
+
+        return CE_None;
+    }
+
+    const std::vector<double> &Results() const
+    {
+        if (BatchCount() == 1)
+        {
+            return m_oExpression->Results();
+        }
+        else
+        {
+            return m_adfResults;
+        }
+    }
+
+  private:
+    int BatchCount() const
+    {
+        return (m_nInBands + m_nNominalBatchSize - 1) / m_nNominalBatchSize;
```

Comment:
- could this be stored in a const variable at construction time to avoid potential recomputation?
- if someone would use an insane large value for m_nNominalBatchSize, the addition could overflow int. You can for example use the DIV_ROUND_UP(m_nInBands, m_nNominalBatchSize) from gcore/gdal_priv.h

Reply:
done, thanks